tcljava/doc/TclBoolean.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: TclBoolean.3,v 1.1 1998/10/14 21:09:08 cvsadmin Exp $
'\" 
.so man.macros
.TH TclBoolean 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclBoolean -- This class is used to create and read boolean
objects from Java code.

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static TclObject \fBnewInstance\fR(boolean \fIbooleanValue\fR)
.sp
static boolean \fBget\fR(Interp \fIinterp\fR, TclObject \fItobj\fR) throws TclException

.SH ARGUMENTS
.AS TclObject "booleanValue" .
.AP boolean booleanValue
New value for internal boolean representation.
.AP Interp interp
If an error occurs while converting an object to be a boolean object,
an error message is left in the interpreter's result object
unless \fIinterp\fR is NULL.
.AP TclObject tobj
TclObject from which to extract a boolean value.  If the internal
representation of \fItobj\fR is not a TclBoolean, the method attempts to
convert it to a TclBoolean.
.BE

.SH DESCRIPTION
.TP
\fBnewInstance\fR
The \fBnewInstance\fR method creates and returns a new TclObject whose
internal representation is the boolean \fIbooleanValue\fR.
.TP
\fBget\fR
The \fBget\fR method returns an boolean value from the \fItobj\fR.  If an error
occurs while converting \fItobj\fR to be a boolean object, an error
message is left in the \fIinterp\fR's result object unless
\fIinterp\fR is NULL.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewBooleanObj, Tcl_GetBooleanFromObj

.SH "SEE ALSO"
InternalRep, TclObject, TclList, TclString, TclInteger, TclDouble, ReflectObject

.SH KEYWORDS
internal representation, object, object type, boolean


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/TclBoolean.java
/*
 * TclBoolean.java
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclBoolean.java,v 1.7 2006/06/07 17:16:10 mdejong Exp $
 *
 */

package tcl.lang;

/**
 * This class implements the boolean object type in Tcl.
 */

public class TclBoolean implements InternalRep {
    /**
     * Internal representations for all TclBooleans.
     */

    private final static TclBoolean trueRep = new TclBoolean(true);
    private final static TclBoolean falseRep = new TclBoolean(false);

    private final boolean value;

    /**
     * Construct a TclBoolean representation with the given boolean value.
     * Note that only two TclBoolean internal rep instances will ever
     * exist.
     *
     * @param b initial boolean value.
     */
    private TclBoolean(boolean b) {
	value = b;
    }

    /**
     * Returns a dupilcate of the current object.
     */
    public InternalRep duplicate() {
	return this;
    }

    /**
     * Implement this no-op for the InternalRep interface.
     */

    public void dispose() {}

    /**
     * Called to query the string representation of the Tcl object. This
     * method is called only by TclObject.toString() when
     * TclObject.stringRep is null.
     *
     * @return the string representation of the Tcl object.
     */
    public String toString() {
	if (value) {
	    return "1";
	} else {
	    return "0";
	}
    }

    /**
     * Creates a new instance of a TclObject with a TclBoolean internal
     * representation.
     *
     * @param b initial value of the boolean object.
     * @return the TclObject with the given boolean value.
     */

    public static TclObject newInstance(boolean b) {
	return new TclObject(b ? trueRep : falseRep);
    }

    /**
     * SetBooleanFromAny -> setBooleanFromAny
     *
     * Called to convert the other object's internal rep to boolean.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to convert to use the
     *     representation provided by this class.
     */
    private static void setBooleanFromAny(Interp interp, TclObject tobj)
	    throws TclException {
	InternalRep rep = tobj.getInternalRep();
	// Get the string representation. Make it up-to-date if necessary.
	String string = tobj.toString();

	if (rep instanceof TclInteger) {
	    int i = TclInteger.get(interp, tobj);
	    if (i == 0) {
	        tobj.setInternalRep(falseRep);
	    } else {
	        tobj.setInternalRep(trueRep);
	    }

	    if (TclObject.saveObjRecords) {
	        String key = "TclInteger -> TclBoolean";
	        Integer num = (Integer) TclObject.objRecordMap.get(key);
	        if (num == null) {
	            num = new Integer(1);
	        } else {
	            num = new Integer(num.intValue() + 1);
	        }
	        TclObject.objRecordMap.put(key, num);
	    }
	} else if (rep instanceof TclDouble) {
	    double d = TclDouble.get(interp, tobj);
	    if (d == 0.0) {
	        tobj.setInternalRep(falseRep);
	    } else {
	        tobj.setInternalRep(trueRep);
	    }

	    if (TclObject.saveObjRecords) {
	        String key = "TclDouble -> TclBoolean";
	        Integer num = (Integer) TclObject.objRecordMap.get(key);
	        if (num == null) {
	            num = new Integer(1);
	        } else {
	            num = new Integer(num.intValue() + 1);
	        }
	        TclObject.objRecordMap.put(key, num);
	    }
	} else {
	    // Copy the string converting its characters to lower case.

	    string = string.toLowerCase();
	    String lowerCase = string.toLowerCase();

	    // Parse the string as a boolean. We use an implementation here that
	    // doesn't report errors in interp if interp is null.

	    boolean b;
	    boolean badBoolean = false;

	    try {
	        b = Util.getBoolean(interp, lowerCase);
	    } catch (TclException te) {
	        // Boolean values can be extracted from ints or doubles.  Note
	        // that we don't use strtoul or strtoull here because we don't
	        // care about what the value is, just whether it is equal to
	        // zero or not.

                badBoolean = true;
                b = false; // Always reassigned below
                if (interp != null) {
                    interp.resetResult();
                }

	        try {
                    b = (Util.getInt(interp, lowerCase) != 0);
                    badBoolean = false;
	        } catch (TclException te2) {}

                if (badBoolean) {
	            try {
                        b = (Util.getDouble(interp, lowerCase) != 0.0);
                        badBoolean = false;
	            } catch (TclException te2) {}
                }
	    }
            if (badBoolean) {
	        if (interp != null) {
	            interp.resetResult();
	        }
	        throw new TclException(interp,
	            "expected boolean value but got \"" +
	            string + "\"");
            }
	    if (b) {
	        tobj.setInternalRep(trueRep);
	    } else {
	        tobj.setInternalRep(falseRep);
	    }

	    if (TclObject.saveObjRecords) {
	        String key = "TclString -> TclBoolean";
	        Integer num = (Integer) TclObject.objRecordMap.get(key);
	        if (num == null) {
	            num = new Integer(1);
	        } else {
	            num = new Integer(num.intValue() + 1);
	        }
	        TclObject.objRecordMap.put(key, num);
	    }
	}
    }

    /**
     * Returns the value of the object as an boolean.
     *
     * An object with a TclBoolean internal rep has
     * a boolean value. An object with a TclInteger
     * internal rep and has the int value "0" or "1"
     * is also a valid boolean value.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to use as an boolean.
     * @return the boolean value of the object.
     * @exception TclException if the object cannot be converted into a
     *     boolean.
     */
    public static boolean get(
	final Interp interp,
	final TclObject tobj)
	    throws TclException
    {
	if (tobj.isIntegerType()) {
	    // An integer with the value 0 or 1 can be
	    // considered a boolean value, so there is
	    // no need to change the internal rep.
	    int ival = tobj.ivalue;  // Inline TclInteger.get()
	    if (ival == 0) {
	        return false;
	    } else if (ival == 1) {
	        return true;
	    }
	}

	InternalRep rep = tobj.getInternalRep();
	TclBoolean tbool;

	if (!(rep instanceof TclBoolean)) {
	    setBooleanFromAny(interp, tobj);
	    tbool = (TclBoolean) tobj.getInternalRep();
	} else {
	    tbool = (TclBoolean) rep;
	}
	return tbool.value;
    }
}

